File: Matlab/calculus/find_root/use_func_fixed_point.m
clc;
clear;
close all;

% Input
a = 1.5;
e = 0.00001;
F = @(x)(x - 1 - 0.3.*cos(x));
g = @(x)(1 + 0.3.*cos(x));

% Algorithm
fprintf("I\ta\t\t\t\tg(a)\t\t\tf(a)\n");
for i = 1:100
    NsolF = g(a);
    NvalF = F(NsolF);
    fprintf("%d)\t%.10f\t%.10f\t%.10f\n", i, a, NsolF, NvalF);
    if abs(NvalF) < e
        break;
    end
    a = NsolF;
end

% Output
fprintf("\nNsolF = %.10f\n", NsolF);

% Compare
EsolF = fzero(F, a);
fprintf("EsolF = %.10f\nError = %.10f\n", EsolF, abs(EsolF - NsolF));

% **************************************************^**************************************************
% *****************************# Copyright by Ali Forouzandeh Hafshejani #*****************************
% ******************************# Scientific Computing Specialist 20@21 #******************************
% **************************************************^**************************************************

File: C/calculus/find_root/use/use_func_fixed_point.c
// Start
#include <math.h>
#include <stdio.h>

// Function
double func(double x)
{
    return (
        x - 1 - 0.3*cos(x)
    );
}

double Gfunc(double x)
{
    return (
        1 + 0.3*cos(x)
    );
}

// Main
int main()
{
    int iter, Niter;
    double ax, err, NsolF, NvalF;

    // Input
    ax = 1.5;
    err = 0.00001;
    Niter = 100;

    // Algorithm
    printf("I\ta\t\tg(a)\t\tf(a)\n");

    for (iter = 0; iter < Niter; ++iter)
    {
        NsolF = Gfunc(ax);
        NvalF = func(NsolF);

        printf("%d)\t%.10f\t%.10f\t%.10f\n",
            iter+1, ax, NsolF, NvalF);

        if (fabs(NvalF) < err)
        {
            break;
        }
        ax = NsolF;
    }

    // Output
    printf("\nNsolF = %.10f\n", NsolF);

    // End
    return 0;
}

// **************************************************^**************************************************
// *****************************# Copyright by Ali Forouzandeh Hafshejani #*****************************
// ******************************# Scientific Computing Specialist 20@21 #******************************
// **************************************************^**************************************************